worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/BukkitAdapter.java

```diff
@@ -387,21 +390,22 @@ public static ItemType asItemType(Material material) {
      */
     public static BlockState adapt(BlockData blockData) {
         checkNotNull(blockData);
-        return blockStateCache.computeIfAbsent(blockData.getAsString(), new Function<String, BlockState>() {
-            @Nullable
-            @Override
-            public BlockState apply(@Nullable String input) {
-                try {
-                    return WorldEdit.getInstance().getBlockFactory().parseFromInput(input, TO_BLOCK_CONTEXT).toImmutableState();
-                } catch (InputParseException e) {
-                    e.printStackTrace();
-                    return null;
-                }
+        int cacheKey =
+                WorldEditPlugin.getInstance().getBukkitImplAdapter() != null
+                        ? WorldEditPlugin.getInstance().getBukkitImplAdapter().getInternalBlockStateId(blockData)
+                            .orElseGet(() -> blockData.getAsString().hashCode())
+                        : blockData.getAsString().hashCode();
```

Comment:
Is there a test we could perform to validate that this doesn't collide? I would prefer to use an incrementing Atomic Integer, but if this can be proven to work its also okay.

Reply:
An incrementing atomic integer wouldn't directly work here, as it needs to be reliably reproducible. As for hash collisions, it theoretically could collide yes. However only when the ImplAdapter is missing